packages/flutter_tools/lib/src/dart/package_map.dart

```diff
@@ -16,6 +16,45 @@ Future<PackageConfig> currentPackageConfig() async {
   return loadPackageConfigUri(Isolate.packageConfigSync!);
 }
 
+/// Locates the `.dart_tool/package_config.json` relevant to [dir].
+///
+/// Searches [dir] and all parent directories.
+///
+/// Returns `null` if no package_config.json was found.
+File? findPackageConfigFile(Directory dir) {
+  final FileSystem fileSystem = dir.fileSystem;
+
+  String candidateDir = fileSystem.path.absolute(dir.path);
+  while (true) {
+    final File candidatePackageConfigFile = fileSystem.file(
+      fileSystem.path.join(candidateDir, '.dart_tool', 'package_config.json'),
+    );
+    if (fileSystem.file(candidatePackageConfigFile).existsSync()) {
+      return candidatePackageConfigFile;
+    }
+    // TODO(sigurdm): we should not need to check this file, it is obsolete,
+    // https://github.com/flutter/flutter/issues/150908.
+    final File candidatePackagesFile = fileSystem.file(fileSystem.path.join(candidateDir, '.packages'));
+    if (fileSystem.file(candidatePackagesFile).existsSync()) {
+      return candidatePackagesFile;
+    }
+    final String parentDir = fileSystem.path.dirname(candidateDir);
+    if (fileSystem.identicalSync(parentDir, candidateDir)) {
+      return null;
+    }
+    candidateDir = parentDir;
+  }
+}
+
+/// Locates the `.dart_tool/package_config.json` relevant to [dir].
+///
+/// Like [findPackageConfigFile] but returns
+/// `$dir/.dart_tool/package_config.json` if no package config could be found.
+File findPackageConfigFileOrDefault(Directory dir) {
```

Comment:
It seems like we call this function a lot, is there any reason we can't cache it as a field in the `Environment` class?

Reply:
Yeah - this is a good point.

If we ensure that it is only called after pub get, then it should be fine.

I guess we are caching it in `BuildInfo` - everywhere should just read it from there.

Added a TODO